### PR TITLE
Allows relative paths for diagnostic scripts.

### LINKED ISSUE #2328: Allow usage of relative paths for diagnostic scripts
Hi everyone,

Currently, in the diagnostic section of a recipe, the script must be either:

- A valid path to an existing diagnostic shipped with ESMValtool (diagnostic root)
- An absolute path to a diagnostic (not portable)

Looking at the code here, there is a third option (not sure that one is very useful though):

https://github.com/ESMValGroup/ESMValCore/blob/6f879cfbff20568ca6b9e2ea4667793a44bef2e5/esmvalcore/_task.py#L350C1-L358C70

It appears that things like `~/my_esmvaltool_diagnostics/my_diagnostic.py` will expand to a valid location.

Absolute paths are not portable...
We are currently developing some material to teach people how to develop their diagnostics.
We would like to be able to ship a `recipe.yml` and a `diagnostic.py` together, with the recipe referencing the diagnostic using a relative path.

I propose to check for a local diagnostic script before appending the esmvaltool diagnostic root to the path.
PR coming.

### PR DESCRIPTION
**Pull request by rbeucher**
## Description

Closes #2328 

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [X] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [X] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [X] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [X] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [X] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [X] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [X] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by valeriupredoi** [comment]
@rbeucher would you be able to run pre-commit on this one please, mate? I'll have a look after - my sincere apologies, no idea why I made this PR wait for so long

**Comment by rbeucher** [comment]
@valeriupredoi , this one should be good to go

File: esmvalcore/_task.py [hunk 1]
```diff
@@ -386,9 +386,15 @@ def _initialize_cmd(self):
...
-        script_file = (diagnostics_root / Path(script).expanduser()).absolute()
 
+        # Check if local diagnostic path exists
+        script_file = Path(script).expanduser().absolute()
         err_msg = f"Cannot execute script '{script}' ({script_file})"
+        if not script_file.is_file():
+            # Try diagnostics_root
+            script_file = (
+                diagnostics_root / Path(script).expanduser()
+            ).absolute()
```
Comment (flicj191) [inline esmvalcore/_task.py:397]: should we add to the error message that another path was tried? 
`err_msg += f" or ({script_file})"`

**Review by flicj191 — COMMENTED** [review]
I have tested and is working for me

**Comment by rbeucher** [comment]
Can you merge @valeriupredoi ?

**Comment by rbeucher** [comment]
@valeriupredoi . It is a relatively minor change. Could you have a look when you get a chance? Thanks!

File: esmvalcore/_task.py (hunk 1 above)
Comment (valeriupredoi) [reply esmvalcore/_task.py:397]: I agree with Felicity here - please add a message saying the custom path didn't work, check it, running diag from base repo etc :beer:

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
many thanks, guys! I think @flicj191 suggestion is a useful add-on; all good otherwise - note however, that the diagnostic will/might fail if the `esmvaltool` package is not installed (this is all done from `esmvalcore`, so maybe add a blocker message saying that)

**Comment by valeriupredoi** [comment]
many thanks @rbeucher and @flicj191  :beer: I made a small correction in c643f558cefba12e4c19db8f9b12692c8194f8fe to change to logging output (INFO and WARN), that corrected the pre-commit issue as well, since `warnings` should be used in a context manager, but I think it's safer to use the logging module here. If you OK with that, please merge at will :+1:

**Review by valeriupredoi — APPROVED** [review]

**Pull request merged**